File: Array/rangeSumQuery2D.cpp
/**
0304 Range Sum Query 2D - Immutable
Prefix Sum
*/

class NumMatrix 
{
public:
    vector<vector<int>> prefixSum;
    NumMatrix(vector<vector<int>>& matrix) 
    {
        int m = matrix.size(), n = matrix[0].size();
        prefixSum.resize(m + 1, vector<int>(n + 1));
        for (int i = 1; i <= m; i++) 
        {
            for (int j = 1; j <= n; j++) 
            {
                prefixSum[i][j] = prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1] + matrix[i - 1][j - 1];   
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) 
    {
        return prefixSum[row2 + 1][col2 + 1] - prefixSum[row1][col2 + 1] - prefixSum[row2 + 1][col1] + prefixSum[row1][col1];
    }
};

File: Array/rangeSumQuery2D.py
'''
0304 Range Sum Query 2D - Immutable
'''

class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        n, m = len(matrix), len(matrix[0])
        self.prefixSum = [[0] * m for _ in range(n)]
        for i in range(n):
            for j in range(m):
                self.prefixSum[i][j] += matrix[i][j]
                if i - 1 > -1:
                    self.prefixSum[i][j] += self.prefixSum[i - 1][j]
                if j - 1 > -1:
                    self.prefixSum[i][j] += self.prefixSum[i][j -1]
                if i - 1 > -1 and j - 1 > -1:
                    self.prefixSum[i][j] -= self.prefixSum[i - 1][j - 1]
        

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        result = self.prefixSum[row2][col2]
        if col1 - 1 > -1:
            result -= self.prefixSum[row2][col1 - 1]
        if row1 - 1 > -1:
            result -= self.prefixSum[row1 - 1][col2]
        if row1 - 1 > -1 and col1 - 1 > -1:
            result += self.prefixSum[row1 - 1][col1 - 1]
        return result
        